pilot/pkg/serviceregistry/kube/controller/endpointslice.go

```diff
@@ -244,29 +245,55 @@ func endpointHealthStatus(svc *model.Service, e v1.Endpoint) model.HealthStatus
 	return model.UnHealthy
 }
 
-func (esc *endpointSliceController) updateEndpointCacheForSlice(hostName host.Name, slice *v1.EndpointSlice) {
+func (esc *endpointSliceController) updateEndpointCacheForSlice(hostName host.Name, epSlice *v1.EndpointSlice) {
 	var endpoints []*model.IstioEndpoint
-	if slice.AddressType == v1.AddressTypeFQDN {
+	if epSlice.AddressType == v1.AddressTypeFQDN {
 		// TODO(https://github.com/istio/istio/issues/34995) support FQDN endpointslice
 		return
 	}
 	svc := esc.c.GetService(hostName)
-	svcNamespacedName := getServiceNamespacedName(slice)
+	svcNamespacedName := getServiceNamespacedName(epSlice)
+	// This is not a endpointslice for service, ignore
+	if svcNamespacedName.Name == "" {
+		return
+	}
 	svcCore := esc.c.services.Get(svcNamespacedName.Name, svcNamespacedName.Namespace)
+	if svcCore != nil && len(svcCore.Spec.ClusterIPs) > 1 {
+		// It means this is dual stack service, for which k8s will build two endpointslice per pod.
+		// If the service is with selector, k8s will manually generate two ip families endpointslices,
+		// we ignore ipv6 family address to prevent generating duplicate IstioEndpoints.
+		if svcCore.Spec.Selector != nil {
```

Comment:
This is not a safe assumption. There is no special logic for service having a selector, they can still have extra endpoints added.

Full example, which works in k8s:

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: shell
spec:
  selector:
    matchLabels:
      app: shell
  template:
    metadata:
      labels:
        app: shell
    spec:
      terminationGracePeriodSeconds: 0
      containers:
      - name: shell
        image: howardjohn/shell
        imagePullPolicy: IfNotPresent
        args:
          - /bin/sleep
          - infinity
        resources:
          requests:
            cpu: 10m
            memory: 128Mi
          limits:
            cpu: 500m
            memory: 500M
---
apiVersion: v1
kind: Service
metadata:
  name: shell
  labels:
    app: shell
spec:
  ports:
  - port: 9087
    name: http
  selector:
    app: shell
---
addressType: IPv4
apiVersion: discovery.k8s.io/v1
endpoints:
- addresses:
  - 1.2.3.4
  conditions:
    ready: true
    serving: true
    terminating: false
kind: EndpointSlice
metadata:
  labels:
    app: shell
    kubernetes.io/service-name: shell
  name: shell-manual
  namespace: default
ports:
- name: http
  port: 9087
  protocol: TCP
```

Reply:
For a service without selector, we  cannot do kind of dedup gracefully.  So ignore the rare case here.

Actuallt a urgly way to do dedup is to iterate over all the endpointslice, and group all the endpoints which belong to a same service. Then for each group, find out whether dual stack endpoints with pod are there.

I think we donot need to process this case, because i think with real pods there, normally people cannot manage the service and endpointslice manually themselves

Reply:
I don't think it's acceptable to incorrectly process endpoints, even if they are "weird" cases. In ambient we initially made assumptions like this and found in real world they ended up breaking real applications. 

It is critical to be correct here or we will cause an outage - and this impacts all users, not just ones with the experiment dual stack mode

Reply:
Do you agree with the above processing or have other suggestion


The above process could have very bad performance

Reply:
I don't think we should have logic around `selector`, just about targetRef. Otherwise this lgtm. Maybe we can fix in a followup?

Reply:
Yes, i kind of agree, with the tests  i can do more investigate